<launch>
  <arg name="uav_name" default="$(optenv UAV_NAME uav2)"/>

  <arg name="standalone" default="false"/>

  <arg unless="$(arg standalone)" name="nodelet" value="load"/>
  <arg     if="$(arg standalone)" name="nodelet" value="standalone"/>
  <arg unless="$(arg standalone)" name="nodelet_manager" value="$(arg uav_name)_uvdar_nodelet_manager"/>
  <arg     if="$(arg standalone)" name="nodelet_manager" value=""/>

  <arg name="threshold" default="100"/>

  <arg name="calibrations_folder" default="$(find mrs_uav_general)/config/uvdar_calibrations"/>


  <!-- Node Settings -->
  <arg name="output" default="screen"/>
  <arg name="proc" default="false"/>
  <arg name="view" default="false"/>
  <arg name="calib" default="false"/>

  <arg name="profiling" default="false"/>
  <arg name="debug" default="false"/>
  <arg name="visual_debug" default="false"/>
  <arg name="gui" default="true"/>
  <arg name="publish" default="true"/>
  
  <arg name="id1" value="0"/>
  <arg name="id2" value="1"/>
  <arg name="id3" value="2"/>
  <arg name="id4" value="3"/>
  <arg name="id5" value="4"/>
  <arg name="id6" value="5"/>
  <arg name="id7" value="6"/>
  <arg name="id8" value="7"/>
  <arg name="id9" value="8"/>
  <arg name="id10" value="9"/>
  <arg name="id11" value="10"/>
  <arg name="id12" value="11"/>

  <arg name="beacon" default="false"/>

  <arg name="sequence_file" default="$(find uvdar_core)/config/selected.txt"/>
  <arg name="enable_manchester" default="false"/>
  <arg name="buffer_size" default="3"/>
  <arg name="model_file" default="$(find uvdar_core)/config/models/quadrotor_foursided.txt"/>
  <!-- ideal angle: 1.221730476 - pi/2 -->

      <!-- <origin xyz="0.108 0 0.085" rpy="1.57079632679 4.71238898038 1.57079632679" /> -->
  <!--Usage: static_transform_publisher x y z yaw pitch roll frame_id child_frame_id  period(milliseconds)-->
<!-- OR -->
  <group ns="$(arg uav_name)">

   <node
    name="uvcam_back_tf_$(arg uav_name)"
    pkg="tf2_ros"
    type="static_transform_publisher"
    args="-0.10 0.0 -0.02 0.0 1.57079632679 3.14159265 $(arg uav_name)/fcu $(arg uav_name)/uvcam_back"/>

    <node pkg="nodelet" type="nodelet" name="$(arg uav_name)_uvdar_nodelet_manager" args="manager" output="screen" >
    <!-- <node pkg="nodelet" type="nodelet" name="$(arg uav_name)_uvdar_nodelet_manager" args="manager" output="screen" launch-prefix="debug_roslaunch"> -->
      <param name="num_worker_threads" value="8" />
    </node>

    <node name="bluefox_emulator" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARBluefoxEmulator $(arg nodelet_manager)" output="screen" respawn="false" launch-prefix="bash -c 'sleep 5; $0 $@'">
      <rosparam param="camera_output_topics"> ["camera_back"] </rosparam>
      <rosparam param="calib_files"> ["default"] </rosparam>
      <remap from="~camera_back" to="/$(arg uav_name)/uvdar_bluefox_back/image_raw"/>
      <remap from="~camera_back_transfer" to="/gazebo/$(arg uav_name)/uvdar_bluefox_back/image_raw"/>
    </node>

    <node name="uv_detect" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARDetector $(arg nodelet_manager)" output="screen" respawn="false"> 
    <!-- <node name="uv_detect" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARDetector $(arg nodelet_manager)" output="screen" respawn="true" launch-prefix="debug_roslaunch"> --> 
      <!-- <node name="uv_detect" pkg="uvdar" type="uv_detector_node" output="screen" launch-prefix="urxvt -e gdb -q -x /home/viktor/gdb.cmds -/-args"> -->
        <!-- <node name="uv_detect" pkg="nodelet" type="nodelet" args="$(arg nodelet) uvdar/UVDARDetector $(arg nodelet_manager)" output="screen" respawn="true" launch-prefix="valgrind -/-tool=callgrind -/-callgrind-out-file=/home/viktor/callgrind.out -/-instr-atstart=no -/-collect-atstart=yes"> --> 

      <param name="uav_name"		type="string" value="$(arg uav_name)"/> 
      <param name="debug"		type="bool" value="$(arg debug)"/> 
      <!-- <param name="gui"             type="bool" value="true"/> -->
      <param name="gui"             type="bool" value="false"/>
      <!-- <param name="publish_visualization"             type="bool" value="true"/> -->
      <param name="publish_visualization"             type="bool" value="false"/>
      <param name="publish"         type="bool" value="$(arg publish)"/>
      <param name="justReport" type="bool" value="true"/>
      <param name="threshold" type="int" value="$(arg threshold)"/>

      <rosparam param="camera_topics"> ["camera_back"] </rosparam>
      <rosparam param="points_seen_topics"> ["points_seen_back"] </rosparam>

      <param name="use_masks"         type="bool" value="false"/>

      <remap from="~camera_back" to="/$(arg uav_name)/uvdar_bluefox_back/image_raw"/>
      <remap from="~points_seen_back" to="/$(arg uav_name)/uvdar/points_seen_back"/>
      <remap from="~points_seen_back/sun" to="/$(arg uav_name)/uvdar/points_seen_back/sun"/>

      <remap from="~odometry" to="/$(arg uav_name)/mrs_odometry/new_odom"/>
      <remap from="~imu" to="mavros/imu/data"/>
    </node>


    
    <node name="UVDAR_BP_Tim" pkg="nodelet" type="nodelet" args="load uvdar/UVDAR_BP_Tim $(arg nodelet_manager)" output="screen" respawn="false">
    <!-- <node name="UVDAR_BP_Tim" pkg="nodelet" type="nodelet" args="load uvdar/UVDAR_BP_Tim $(arg nodelet_manager)" output="screen" respawn="false" launch-prefix="debug_roslaunch"> -->

      <param name="uav_name" type = "string" value="$(arg uav_name)"/>
      <param name="debug"		type="bool" value="$(arg debug)"/> 
      <param name="visual_debug"		type="bool" value="$(arg visual_debug)"/> 
      <param name="gui"             type="bool" value="$(arg gui)"/>
      
      <!-- Changeable for each nodelet independently. - no arguments! -->
      <param name="publish_visualization"             type="bool" value="false"/>
      <param name="use_camera_for_visualization"         type="bool" value="false"/>
      <param name="visualization_rate"         type="double" value="2.0"/>


      <param name="sequence_file"         type="string" value="$(arg sequence_file)"/>
      <param name="enable_manchester"     type="bool"   value="$(arg enable_manchester)"/>
      <param name="buffer_size"           type="int"    value="$(arg buffer_size)"/>

      <rosparam param="camera_topics"> ["camera_back"] </rosparam>
      <rosparam param="points_seen_topics"> ["points_seen_back"] </rosparam>
      <rosparam param="blinkers_seen_topics"> ["blinkers_seen_back"] </rosparam>
      <rosparam param="estimated_framerate_topics"> ["estimated_framerate_back"] </rosparam>

      <remap from="~camera_back" to="/$(arg uav_name)/uvdar_bluefox_back/image_raw"/>
      <remap from="~points_seen_back" to="/$(arg uav_name)/uvdar/points_seen_back"/>
      <remap from="~blinkers_seen_back" to="/$(arg uav_name)/uvdar/blinkers_seen_back"/>
      <remap from="~estimated_framerate_back" to="/$(arg uav_name)/uvdar/estimated_framerate_back"/>
      <remap from="~visualization" to="/$(arg uav_name)/uvdar/blink_visualization/image_raw"/>

    </node>
      
     <node name="uvdar_pose_calculator_node" pkg="uvdar_core" type="uvdar_pose_calculator_node" output="screen">
    <!-- launch-prefix="debug_roslaunch"> -->
    <!-- <node name="uvdar_pose_calculator_node" pkg="uvdar_core" type="uvdar_pose_calculator_node" output="screen" > --> 
        <!-- <node name="uvdar_pose_calculator_node" pkg="uvdar_core" type="uvdar_pose_calculator_node" launch-prefix="valgrind -/-tool=callgrind -/-callgrind-out-file=/home/viktor/callgrind.out -/-instr-atstart=no -/-collect-atstart=yes"> --> 
            <!-- <node name="uvdar_pose_calculator_node" pkg="uvdar_core" type="uvdar_pose_calculator_node" launch-prefix="valgrind -/-tool=callgrind -/-callgrind-out-file=/home/viktor/callgrind.out"> --> 
      <param name="uav_name" type = "string" value="$(arg uav_name)"/>
      <!-- <param name="debug"		type="bool" value="true"/> --> 
      <param name="debug"		type="bool" value="false"/> 
      <!-- <param name="gui"             type="bool" value="false"/> -->
      <param name="gui"             type="bool" value="false"/>
      <param name="publish_visualization"             type="bool" value="false"/>
      <param name="publish_constituents"             type="bool" value="true"/>
      <!-- <param name="publish_visualization"             type="bool" value="false"/> -->
      <!-- <param name="frequencies_per_target" type="int" value="1"/> -->
      <!-- <rosparam param="frequencies" subst_value="true"> [$(arg frequency1), $(arg frequency2), $(arg frequency3)] </rosparam> -->
      <rosparam param="signal_ids" subst_value="true"> [$(arg id1), $(arg id2), $(arg id3), $(arg id4), $(arg id5), $(arg id6), $(arg id7), $(arg id8), $(arg id9), $(arg id10), $(arg id11), $(arg id12)] </rosparam>


      <param name="profiling" type="bool" value="$(arg profiling)"/>

      <param name="quadrotor" type="bool" value="false"/>
      <param name="custom_model" type="bool" value="true"/>
      <param name="model_file" type="string" value="$(arg model_file)"/>
      <param name="beacon" type="bool" value="false"/>

      <rosparam param="blinkers_seen_topics"> ["blinkers_seen_back"] </rosparam>
      <rosparam param="estimated_framerate_topics"> ["estimated_framerate_back"] </rosparam>
      <rosparam param="camera_frames" subst_value="true"> ["$(arg uav_name)/uvcam_back"] </rosparam>

      <remap from="~blinkers_seen_back" to="/$(arg uav_name)/uvdar/blinkers_seen_back"/>
      <remap from="~estimated_framerate_back" to="/$(arg uav_name)/uvdar/estimated_framerate_back"/>

      <rosparam param="calib_files" subst_value="True"> ["default", "default", "default"] </rosparam>

      <remap from="~constituentPoses1" to="/$(arg uav_name)/uvdar/constituentPosesL"/>
      <remap from="~constituentPoses2" to="/$(arg uav_name)/uvdar/constituentPosesR"/>
      <remap from="~constituentPoses3" to="/$(arg uav_name)/uvdar/constituentPosesB"/>

      <remap from="~measuredPoses1" to="/$(arg uav_name)/uvdar/measuredPosesL"/>
      <remap from="~measuredPoses2" to="/$(arg uav_name)/uvdar/measuredPosesR"/>
      <remap from="~measuredPoses3" to="/$(arg uav_name)/uvdar/measuredPosesB"/>
    </node>
  </group>


</launch>
